### DISCUSSION TITLE
How to avoid escape characters in links with `remark-gfm` and `strip-markdown`?

### DISCUSSION THREAD

**Discussion by stefanprobst** [remarkjs/remark#963] [Q&A] [answered]
when using both `remark-gfm` and `strip-markdown`, some characters are being escaped (seems to happen [here](https://github.com/syntax-tree/mdast-util-to-markdown/blob/8901ff34c2c1346f10fd119142fd8d78b3f52827/lib/util/safe.js#L101) in `mdast-util-to-markdown`)

how can i avoid this?

example:

```js
import { remark } from "remark";
import gfm from "remark-gfm";
import strip from "strip-markdown";

const processor = remark().use(gfm).use(strip);
const result = processor.processSync("http://www.example.com");

console.log(String(result));
// prints: http\://www\.example.com
// expected: http://www.example.com
```

**Comment by ChristianMurphy** [comment, marked as answer]
Have you tried using https://github.com/syntax-tree/mdast-util-to-string as a compiler?
the default compiler uses https://github.com/syntax-tree/mdast-util-to-markdown which will always do escaping for markdown-like-but-not-markdown constructs.
https://github.com/remarkjs/remark/blob/5730ad2170e7ab7797a687757191da984f33415d/packages/remark-stringify/lib/index.js#L7

**Reply by stefanprobst** [reply]
perfect, thanks! the following works!

```js
import { unified } from "unified";
import fromMarkdown from "remark-parse";
import gfm from "remark-gfm";
import strip from "strip-markdown";
import { toMarkdown } from "mdast-util-to-markdown";

const processor = unified()
  .use(fromMarkdown)
  .use(gfm)
  .use(strip)
  .use(function () {
    this.Compiler = toMarkdown;
  });
const result = processor.processSync("http://www.example.com");

console.log(String(result));
// http://www.example.com
```

**Reply by wooorm** [reply]
Uhh, nope. That should not work for other things?

**Reply by stefanprobst** [reply]
> That should not work for other things?

can you clarify? so far i have only tried with simple stuff like this:

```
const result = processor.processSync(
  "This *is* a link: http://www.example.com."
);
console.log(String(result));
// This is a link: http://www.example.com.
```

**Reply by ChristianMurphy** [reply]
@wooorm could you expand on what you mean?
Markdown syntax was and is being removed by `strip-markdown`, there may be some overlap between that and `mdast-util-to-string` and `strip-markdown`, but it should work?

**Reply by wooorm** [reply]
You’re dropping some character references. Not all. This is not intended to work at all, and is not what Christian suggested!

With `this.Compiler = toMarkdown`, you’re overwriting the code that was:

https://github.com/remarkjs/remark/blob/5730ad2170e7ab7797a687757191da984f33415d/packages/remark-stringify/lib/index.js#L12-L28

Namely, GFM extensions don’t work anymore. As in, tables don’t compile (though `strip-markdown` removed those). And because those don‘t compile, URLs also don’t have to be encoded anymore.

**Reply by wooorm** [reply]
@ChristianMurphy You suggested using `toString`. This uses `toMarkdown`. As a `Compiler` (which passes a `VFile` as `options` to `toMarkdown`).

**Reply by stefanprobst** [reply]
sorry, i misread the suggestion.

**Reply by ChristianMurphy** [reply]
Okay?
I'm still a bit lost, if all the markdown syntax is being stripped, then it shouldn't matter if some (unused) AST node compilers are missing or if it is passed as a `VFile`?

**Reply by stefanprobst** [reply]
so, the suggestion would be the following, correct?

```js
const processor = unified().use(fromMarkdown).use(gfm).use(strip);
const tree = processor.parse("This *is* a ~~link~~: http://www.example.com.");
console.log(String(toString(tree)));
// This is a link: http://www.example.com.
```

**Reply by wooorm** [reply]
The `Compiler` signature is `(tree: Node, file: VFile) -> any` (where `any` is typically a `string`).
The signature of `fromMarkdown` is `(tree: Node, options?: Options) -> string`.
The signature of `toString` is `(tree: Node, options?: Options) -> string`.
They are not the same. `fromMarkdown` and `toString` are not a compiler.

Christian suggested `toString`. That would mean, I believe:

```js
const processor = unified()
  .use(fromMarkdown)
  .use(gfm)
  .use(strip)
  .use(function () {
    this.Compiler = (tree) => toString(tree)
  });
```

It’s a good solution but that would not support blank lines between paragraphs though.

`fromMarkdown` *can* somewhat work, but to prevent bugs the signature should be:

```js
const processor = unified()
  .use(fromMarkdown)
  .use(gfm)
  .use(strip)
  .use(function () {
    this.Compiler = (tree) => fromMarkdown(tree)
  });
```

The above code does have a clean signature, and does solve this question.
But it’s not needed, because `remark-stringify` already uses `fromMarkdown`. A cleaner solution in my opinion would not repeat code.
And it’s not perfect: there are still other character references that are used by `fromMarkdown`. I’m guessing that @stefanprobst also does not want those.

**Reply by stefanprobst** [reply]
thank you both!

**Reply by wooorm** [reply]
So, do you care about those other character references, and cleaning it up? Or is this good enough?

**Reply by stefanprobst** [reply]
afaict this is good enough, thanks

**Comment by wooorm** [comment]
See https://github.com/remarkjs/strip-markdown/issues/13 and other issues.

I’m open to rewriting that project to add support for not generating character references at all. It’s not high on my priority list though.
